# **Maestría en Inteligencia Artificial Aplicada**

## Curso: **Procesamiento de Lenguaje Natural**

### Tecnológico de Monterrey

### Prof Luis Eduardo Falcón Morales

## Adtividad Semana 5

### **Vectores Embebidos de OpenAI**

#### **Nombres y matrículas de los integrantes del equipo:**


* Fernando Omar Salazar Ortiz - A01796214
* Carlos Aaron Bocanegra Buitron - A01796345
* Luis Enrique González González - A01795338
* Gloria María Campos García - A01422345



In [5]:
!git clone https://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@github.com/luisegg777/nlp-team-12-aj-2025-tec.git

Cloning into 'nlp-team-12-aj-2025-tec'...


In [8]:
%cd nlp-team-12-aj-2025-tec/

/content/nlp-team-12-aj-2025-tec


In [14]:
!ls /content/drive/MyDrive/Colab\ Notebooks/NLP/Actividades/ActividadSemana5_WordEmbeddings/Equipo12_semana05Embeddings_luis.ipynb

'/content/drive/MyDrive/Colab Notebooks/NLP/Actividades/ActividadSemana5_WordEmbeddings/Equipo12_semana05Embeddings_luis.ipynb'


In [ ]:
!ls /content/drive/MyDrive/Colab\ Notebooks/NLP/Actividades/ActividadSemana5_WordEmbeddings/Equipo12_semana05Embeddings_luis.ipynb .

In [10]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# Aquí deberás incluir todas las librerías que requieras durante esta actividad:

import pandas as pd
import re
import nltk
from nltk.corpus import stopwords
nltk.download('punkt_tab')
nltk.download('stopwords')    # para tener acceso a "stopwords" en varios idiomas.

import spacy #Para aplicar la lemmatization
nlp = spacy.load("en_core_web_sm")

from sklearn.model_selection import train_test_split

from collections import Counter

from google.colab import userdata
from openai import OpenAI
import openai

import pickle

import numpy as np

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV

from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [ ]:
# Incluye las celdas necesarias para tu acceso a la API de OpenAI.

api_key = userdata.get("OPENAI_API_KEY_PERSONAL")

if not api_key:
  raise ValueError("API key no encontrada en los secretos")

client = OpenAI(api_key = api_key)

# **Pregunta - 1:**



Descarga los 3 archivos de Canvas y genera un nuevo DataFrame de Pandas con ellos.

**Llama simplemente "df" a dicho DataFrame.**

Los archivos los encuentras en Canvas: amazon5.txt, imdb5.txt, yelp5.txt.



In [ ]:

# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********


dfa = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/NLP/Actividades/ActividadSemana5_WordEmbeddings/amazon5.txt', sep='\t', names=['review','label'], header=None, encoding='utf-8')
dfi = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/NLP/Actividades/ActividadSemana5_WordEmbeddings/imdb5.txt', sep=r'\s{3,}', names=['review','label'], header=None, encoding='utf-8', engine='python')
dfy = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/NLP/Actividades/ActividadSemana5_WordEmbeddings/yelp5.txt', sep='\t', names=['review','label'], header=None, encoding='utf-8')

df = pd.concat([dfa, dfi, dfy], ignore_index=True)

# *********** Aquí termina la sección de agregar código *************


In [ ]:
# Verifiquemos la información del DataFrame:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   review  3000 non-null   object
 1   label   3000 non-null   int64 
dtypes: int64(1), object(1)
memory usage: 47.0+ KB


In [ ]:
# Y veamos sus primeros registros:
df.head()

review  label
0  So there is no way for me to plug it in here i...      0
1                        Good case, Excellent value.      1
2                             Great for the jawbone.      1
3  Tied to charger for conversations lasting more...      0
4                                  The mic is great.      1

# **Pregunta - 2:**

In [ ]:
stopwords.words('english')

['a',
 'about',
 'above',
 'after',
 'again',
 'against',
 'ain',
 'all',
 'am',
 'an',
 'and',
 'any',
 'are',
 'aren',
 "aren't",
 'as',
 'at',
 'be',
 'because',
 'been',
 'before',
 'being',
 'below',
 'between',
 'both',
 'but',
 'by',
 'can',
 'couldn',
 "couldn't",
 'd',
 'did',
 'didn',
 "didn't",
 'do',
 'does',
 'doesn',
 "doesn't",
 'doing',
 'don',
 "don't",
 'down',
 'during',
 'each',
 'few',
 'for',
 'from',
 'further',
 'had',
 'hadn',
 "hadn't",
 'has',
 'hasn',
 "hasn't",
 'have',
 'haven',
 "haven't",
 'having',
 'he',
 "he'd",
 "he'll",
 'her',
 'here',
 'hers',
 'herself',
 "he's",
 'him',
 'himself',
 'his',
 'how',
 'i',
 "i'd",
 'if',
 "i'll",
 "i'm",
 'in',
 'into',
 'is',
 'isn',
 "isn't",
 'it',
 "it'd",
 "it'll",
 "it's",
 'its',
 'itself',
 "i've",
 'just',
 'll',
 'm',
 'ma',
 'me',
 'mightn',
 "mightn't",
 'more',
 'most',
 'mustn',
 "mustn't",
 'my',
 'myself',
 'needn',
 "needn't",
 'no',
 'nor',
 'not',
 'now',
 'o',
 'of',
 'off',
 'on',
 'once',
 'on

Realiza el proceso de limpieza. Aplica el preprocesamiento que consideres adecuado.











In [ ]:

# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********

def clean_tok(doc):

  #Convertir a minusculas
  docTmp = doc.lower()

  #Sustitucion de negaciones estilo don't por do not

  contractions = {
    r"don't": "do not",
    r"dont": "do not",
    r"doesn't": "does not",
    r"didn't": "did not",
    r"can't": "can not",
    r"cant": "can not",
    r"couldn't": "could not",
    r"won't": "will not",
    r"wouldn't": "would not",
    r"shouldn't": "should not",
    r"mustn't": "must not",
    r"mightn't": "might not",
    r"needn't": "need not",
    r"shan't": "shall not",
    r"hasn't": "has not",
    r"haven't": "have not",
    r"hadn't": "had not",
    r"wasn't": "was not",
    r"weren't": "were not",
    r"isn't": "is not",
    r"aren't": "are not",
  }

  for pattern, replacement in contractions.items():
    docTmp = re.sub(pattern, replacement, docTmp)


  #Reemplazar todos los caracteres no alfabeticos por espacios en blanco
  docTmp2 = re.sub(r'[^a-z]', ' ', docTmp)


  #Tokenizacion usando word_tokenize
  tokensTmp = nltk.word_tokenize(docTmp2)

  #Remover stopwords en ingles
  # Consideremos la siguiente lista de palabras asociada a negaciones en ingles
  negwords = [ 'no', 'nor', 'not', 'ain', 'aren', "aren't", 'don', "don't", 'couldn', "couldn't", 'didn', "didn't", 'doesn', "doesn't", 'hadn', "hadn't", 'hasn', "hasn't", 'haven', "haven't", 'isn', "isn't", 'mightn', "mightn't", 'mustn', "mustn't", 'needn', "needn't", 'shan', "shan't", 'shouldn', "shouldn't", 'wasn', "wasn't", 'weren', "weren't", 'won', "won't", 'wouldn', "wouldn't"]
  mystopwords = [stopword for stopword in stopwords.words('english') if stopword not in negwords]
  tokensTmp2 = [token for token in tokensTmp if token not in mystopwords]

  tokens = tokensTmp2

  return tokens


def clean_doc(doc):
  tokens = []

  #procesamiento de SO
  tmp = [re.sub(r'^[s]+[o]{2,}$', 'so', token) for token in doc]

  #procesamiento de GOOD
  tmp2 = [re.sub(r'^[g]+[o]{3,}[d]+$', 'good', token) for token in tmp]

  #Reemplazar letras repetidas 3+ por una letra
  tmp3 = [re.sub(r'(.)\1{2,}', r'\1', token) for token in tmp2]

  #Aplicar Lemmatization
  tmp4 = [nlp(token)[0].lemma_ for token in tmp3]

  #tokens = tmp3
  tokens = tmp4

  #print(tokens)
  return tokens



# Separamos la información:
#     La "X" serán los datos de entrada, los comentarios.
#     La "Y" será la variable de salida, la evaluación.
X = df.review
Y = df.label



Xcleantok = [clean_tok(x) for x in X]

Xclean = [clean_doc(x) for x in Xcleantok]

# *********** Aquí termina la sección de agregar código *************

In [ ]:
# Despleguemos los primeros comentarios después de tu proceso de limpieza:

# comment 27 had 'didn't'

#for x in Xclean[27:28]:
for x in Xclean[0:5]:
  print(x)


['no', 'way', 'plug', 'we', 'unless', 'go', 'converter']
['good', 'case', 'excellent', 'value']
['great', 'jawbone']
['tie', 'charger', 'conversation', 'last', 'minute', 'major', 'problem']
['mic', 'great']


# **Pregunta - 3:**



Realicemos una partición aleatoria con los mismos porcentajes de la práctica pasada para poder comparar dichos resultados con los de
esta actividad, a saber, 70%, 15% y 15%, para entrenamiento, validación y prueba, respectivamente.

In [ ]:

# ************* Inicia la sección de agregar código:*****************************

x_train, x_val_and_test, y_train, y_val_and_test = train_test_split(Xclean, Y, train_size=.70, shuffle=True, random_state=1)
x_val, x_test, y_val, y_test = train_test_split(x_val_and_test, y_val_and_test, test_size=.50, shuffle=True, random_state=17)


# *********** Termina la sección de agregar código *************


# verificemos las dimensiones obtenidas:
print('X,y Train:', len(x_train), len(y_train))
print('X,y Val:', len(x_val), len(y_val))
print('X,y Test', len(x_test), len(y_test))

X,y Train: 2100 2100
X,y Val: 450 450
X,y Test 450 450


# **Pregunta - 4:**



Construye tu vocabulario a continuación


In [ ]:
# a.	Usa el conjunto de entrenamiento para generar tu vocabulario
#     con un tamaño que consideres adecuado:


# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********

def get_dictionary(comments, min_freq=2, min_word_length=2):
  midicc = {}

  #Se eliminan los tokens que no cumplen con min_word_length
  commentsTmp = [[token for token in comment if len(token) >= min_word_length] for comment in comments]

  #Se calculan las frecuencias de los tokens
  for comment in commentsTmp:
    for token in comment:
      if token in midicc:
        midicc[token] += 1
      else:
        midicc[token] = 1


  #Elimina palabras que no tienen la minima frecuencia
  for token in list(midicc):
    if midicc[token] < min_freq:
      del midicc[token]

  return midicc


midicc = get_dictionary(comments=x_train, min_freq=1, min_word_length=2)
#print(dictionary)



# *********** Aquí termina la sección de agregar código *************

In [ ]:
# b.	Indica el tamaño del vocabulario generado.

print('Longitud del vocabulario generado:')


# ******* Inicia la sección de agregar código: ***********


print(len(midicc))


# *********** Aquí termina la sección de agregar código *************

Longitud del vocabulario generado:
3261


c.	¿Por qué debe usarse solamente el conjunto de entrenamiento para generar el vocabulario?


### ++++++++ Inicia la sección de agregar texto: +++++++++++

**Porque así se evita el filtrado de información**

### ++++++++ Termina la sección de agregar texto: +++++++++++


In [ ]:
# d.	Con el vocabulario generado, filtra los conjuntos de entrenamiento,
#     validación y prueba para que todos los comentarios usen solamente las
#     palabras de este vocabulario.

#     Llamar train_x, val_x y test_x a estos tres conjuntos.


# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********

train_x = []
for ss in x_train:
  train_x.append([w for w in ss if w in midicc])

val_x = []
for ss in x_val:
  val_x.append([w for w in ss if w in midicc])

test_x = []
for ss in x_test:
  test_x.append([w for w in ss if w in midicc])


# *********** Aquí termina la sección de agregar código *************


In [ ]:
# Vemos el resultado de los primeros comentarios del conjunto de entrenamiento:

for ss in train_x[0:5]:
  print(ss)

['co', 'star', 'not', 'fare', 'much', 'well', 'people', 'like', 'morgan', 'freeman', 'jonah', 'hill', 'ed', 'helm', 'waste']
['tonight', 'elk', 'filet', 'special', 'suck']
['pay', 'bill', 'not', 'tip', 'feel', 'server', 'terrible', 'job']
['call', 'steakhouse', 'not', 'properly', 'cook', 'steak', 'not', 'understand']
['however', 'keypad', 'tinny', 'sometimes', 'reach', 'wrong', 'button']


# **Pregunta - 5:**


#### **Incluye aquí un resumen de las características y diferencias que tiene al menos los tres modelos de OpenAI indicados: "text-embedding-3-small", "text-embedding-3-large" y "text-embedding-ada-002".**

### ++++++++ Inicia la sección de agregar texto: +++++++++++

** A continuación se muestra una tabla resumen comparativa de las cracterísticas de los tres modelos:**

| Caracteristica | text-embedding-ada-002 | text-embedding-3-small | text-embedding-3-large   |
|--------|--------|------|--------|
| Rendimiento   | Bajo   | Medio   | Alto |
| Velocidad   | Promedio    | Alta   | Promedio |
| Precio (USD)   | 0.1 por 1M tokens    | 0.02 por 1M tokens   | 0.13 por 1M tokens    |

### ++++++++ Termina la sección de agregar texto: +++++++++++


# **Pregunta - 6:**


#### **Diccionario clave-valor de palabras del diccionario y vectores embebidos.**

In [ ]:
'''
# To test the embedding endpoint: with a single word
response = client.embeddings.create(
    input="DOG",
    model="text-embedding-3-small"
)

print(response.data[0].embedding)
'''

'\n# To test the embedding endpoint: with a single word\nresponse = client.embeddings.create(\n    input="DOG",\n    model="text-embedding-3-small"\n)\n\nprint(response.data[0].embedding)\n'

In [ ]:
#len(response.data[0].embedding)

In [ ]:
'''
from sentence_transformers import SentenceTransformer

model = SentenceTransformer('all-MiniLM-L6-v2')  # Modelo ligero, rápido y preciso
'''



"\nfrom sentence_transformers import SentenceTransformer\n\nmodel = SentenceTransformer('all-MiniLM-L6-v2')  # Modelo ligero, rápido y preciso\n"

In [ ]:
'''
new_dictionary = {}

for token in midicc:
  new_dictionary[token] = model.encode(token)
'''

'\nnew_dictionary = {}\n\nfor token in midicc:\n  new_dictionary[token] = model.encode(token)\n'

In [ ]:
# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********

new_dictionary = {}

'''
for token in midicc:
  #print(token)
  response = client.embeddings.create(
      input=token,
      model="text-embedding-3-small"
  )
  new_dictionary[token] = response.data[0].embedding
'''

# *********** Aquí termina la sección de agregar código *************

'\nfor token in midicc:\n  #print(token)\n  response = client.embeddings.create(\n      input=token,\n      model="text-embedding-3-small"\n  )\n  new_dictionary[token] = response.data[0].embedding\n'

In [ ]:
#len(new_dictionary['star'])


In [ ]:
#new_dictionary['star']

In [ ]:
diccionario_file_name = 'new_dictionary_small.pkl'

In [ ]:
'''
with open('/content/drive/MyDrive/Colab Notebooks/NLP/Actividades/ActividadSemana5_WordEmbeddings/'+diccionario_file_name, 'wb') as f:
  pickle.dump(new_dictionary, f)
'''

"\nwith open('/content/drive/MyDrive/Colab Notebooks/NLP/Actividades/ActividadSemana5_WordEmbeddings/'+diccionario_file_name, 'wb') as f:\n  pickle.dump(new_dictionary, f)\n"

**Los Tokens que se usaron de OpenAI al obtener los vectores de las 3261 palabras del diccionario fueron 5581 con el modelo: text-embedding-3-small (la imagen muestra 5581 porque hice un par de pruebas antes que usaron 2 tokens)**

# **Pregunta - 7:**



Generamos los vectores embebidos a partir de los conjuntos de entrenamiento, validación y prueba.

Los llamaremos trainEmb, valEmb y testEmb, respectivamente.

In [ ]:
'''
count = 0
for comment in train_x:
  if comment != []:
    count += 1

print(count)
'''

'\ncount = 0\nfor comment in train_x:\n  if comment != []:\n    count += 1\n\nprint(count)\n'

In [ ]:
#len([comment for comment in train_x if len(comment) == 1])

In [ ]:
#train_x[7:9]
#[comment for comment in train_x if len(comment) == 1]

In [ ]:
#Se leen los vectores
with open('/content/drive/MyDrive/Colab Notebooks/NLP/Actividades/ActividadSemana5_WordEmbeddings/'+diccionario_file_name, 'rb') as f:
    vectores_midicc_guardados = pickle.load(f)

In [ ]:
len(vectores_midicc_guardados)

3261

In [ ]:
len(vectores_midicc_guardados['star'])

1536

In [ ]:
# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********

dimensiones_vectores = 1536

trainEmb = []
for comment in train_x:
    vectores = [vectores_midicc_guardados[token] for token in comment if token in vectores_midicc_guardados]

    if vectores:
        mean_vec = np.mean(vectores, axis=0)
    else:
        mean_vec = np.zeros(dimensiones_vectores)

    trainEmb.append(mean_vec)

# Se convierte la lista a un array de Numpy para poder leer el atributo shape
trainEmb = np.array(trainEmb)

valEmb = []
for comment in val_x:
    vectores = [vectores_midicc_guardados[token] for token in comment if token in vectores_midicc_guardados]

    if vectores:
        mean_vec = np.mean(vectores, axis=0)
    else:
        mean_vec = np.zeros(dimensiones_vectores)

    valEmb.append(mean_vec)

# Se convierte la lista a un array de Numpy para poder leer el atributo shape
valEmb = np.array(valEmb)

testEmb = []
for comment in test_x:
    vectores = [vectores_midicc_guardados[token] for token in comment if token in vectores_midicc_guardados]

    if vectores:
        mean_vec = np.mean(vectores, axis=0)
    else:
        mean_vec = np.zeros(dimensiones_vectores)

    testEmb.append(mean_vec)

# Se convierte la lista a un array de Numpy para poder leer el atributo shape
testEmb = np.array(testEmb)


# *********** Aquí termina la sección de agregar código *************

In [ ]:
# Veamos las dimensiones de cada conjunto embebido:

print("Train-Emb:", trainEmb.shape)
print("Val-Emb:", valEmb.shape)
print("Test-Emb:", testEmb.shape)

Train-Emb: (2100, 1536)
Val-Emb: (450, 1536)
Test-Emb: (450, 1536)


# **Pregunta - 8:**



Utiliza los modelos de regresión logística y bosque aleatorio (random forest) y encuentra sus desempeños.

Compara los resultados con los de la semana anterior.

In [ ]:
solver_list = ['lbfgs', 'liblinear']
penalty_list = ['l1', 'l2']
max_iter_list = [50, 100]
C_list = [0.1, 0.5, 1.0, 2.0, 4.0, 6.0, 8.0, 10.0]
solver_penalty_combinatons = ['lbfgs-l2', 'lbfgs-None', 'liblinear-l1', 'liblinear-l2']
resultados = []

for a in range(len(solver_list)):
  for b in range(len(penalty_list)):
    for c in range(len(max_iter_list)):
      for d in range(len(C_list)):
        if(solver_list[a]+'-'+penalty_list[b] in solver_penalty_combinatons):
          modeloLREmb = LogisticRegression(solver=solver_list[a], penalty=penalty_list[b], max_iter=max_iter_list[c], C=C_list[d])
          modeloLREmb.fit(trainEmb, y_train)

          resultados.append({
              'solver': solver_list[a],
              'penalty': penalty_list[b],
              'max_iter': max_iter_list[c],
              'C': C_list[d],
              'train_accuracy': (100*modeloLREmb.score(trainEmb, y_train)),
              'val_accuracy': (100*modeloLREmb.score(valEmb, y_val)),
              'train_accuracy - val_accuracy': (100*modeloLREmb.score(trainEmb, y_train)) - (100*modeloLREmb.score(valEmb, y_val))
          })

df_resultados = pd.DataFrame(resultados)

df_resultados



solver penalty  max_iter     C  train_accuracy  val_accuracy  \
0       lbfgs      l2        50   0.1       85.000000     85.777778   
1       lbfgs      l2        50   0.5       85.952381     86.444444   
2       lbfgs      l2        50   1.0       86.476190     86.888889   
3       lbfgs      l2        50   2.0       87.047619     88.000000   
4       lbfgs      l2        50   4.0       88.285714     88.222222   
5       lbfgs      l2        50   6.0       88.333333     88.222222   
6       lbfgs      l2        50   8.0       88.714286     88.000000   
7       lbfgs      l2        50  10.0       88.761905     87.777778   
8       lbfgs      l2       100   0.1       85.000000     85.777778   
9       lbfgs      l2       100   0.5       85.952381     86.444444   
10      lbfgs      l2       100   1.0       86.476190     86.888889   
11      lbfgs      l2       100   2.0       87.047619     88.000000   
12      lbfgs      l2       100   4.0       88.285714     88.222222   
13      lbfgs      l2       100   6.0       88.333333     88.222222   
14      lbfgs      l2       100   8.0       88.714286     88.000000   
15      lbfgs      l2       100  10.0       88.761905     87.777778   
16  liblinear      l1        50   0.1       50.857143     48.000000   
17  liblinear      l1        50   0.5       80.666667     81.777778   
18  liblinear      l1        50   1.0       84.095238     85.333333   
19  liblinear      l1        50   2.0       86.476190     85.777778   
20  liblinear      l1        50   4.0       88.333333     86.000000   
21  liblinear      l1        50   6.0       89.380952     85.777778   
22  liblinear      l1        50   8.0       89.952381     85.555556   
23  liblinear      l1        50  10.0       90.666667     86.000000   
24  liblinear      l1       100   0.1       50.857143     48.000000   
25  liblinear      l1       100   0.5       80.666667     81.777778   
26  liblinear      l1       100   1.0       84.095238     85.333333   
27  liblinear      l1       100   2.0       86.571429     85.777778   
28  liblinear      l1       100   4.0       88.333333     86.000000   
29  liblinear      l1       100   6.0       89.380952     85.777778   
30  liblinear      l1       100   8.0       89.904762     85.777778   
31  liblinear      l1       100  10.0       90.666667     86.000000   
32  liblinear      l2        50   0.1       84.904762     85.555556   
33  liblinear      l2        50   0.5       85.952381     86.444444   
34  liblinear      l2        50   1.0       86.523810     86.888889   
35  liblinear      l2        50   2.0       87.571429     87.555556   
36  liblinear      l2        50   4.0       88.190476     88.444444   
37  liblinear      l2        50   6.0       88.523810     88.000000   
38  liblinear      l2        50   8.0       88.619048     88.222222   
39  liblinear      l2        50  10.0       88.904762     88.000000   
40  liblinear      l2       100   0.1       84.904762     85.555556   
41  liblinear      l2       100   0.5       85.952381     86.444444   
42  liblinear      l2       100   1.0       86.523810     86.888889   
43  liblinear      l2       100   2.0       87.571429     87.555556   
44  liblinear      l2       100   4.0       88.190476     88.444444   
45  liblinear      l2       100   6.0       88.523810     88.000000   
46  liblinear      l2       100   8.0       88.619048     88.222222   
47  liblinear      l2       100  10.0       88.904762     88.000000   

    train_accuracy - val_accuracy  
0                       -0.777778  
1                       -0.492063  
2                       -0.412698  
3                       -0.952381  
4                        0.063492  
5                        0.111111  
6                        0.714286  
7                        0.984127  
8                       -0.777778  
9                       -0.492063  
10                      -0.412698  
11                      -0.952381  
12                       0.063492  
13                     

In [ ]:
### GridSearchCV para la busqueda hiperparametros de LR ########

'''
param_grid = [
    {
        'solver': ['liblinear'],
        'penalty': ['l1', 'l2'],
        'C': [0.1, 0.5, 1.0, 2.0, 4.0],
        'max_iter': [50, 100]
    },
    {
        'solver': ['lbfgs'],
        'penalty': ['l2'],
        'C': [0.1, 0.5, 1.0, 2.0, 4.0],
        'max_iter': [50, 100]
    }
]


modeloLREmb2 = LogisticRegression()
grid_search = GridSearchCV(estimator=modeloLREmb2, param_grid=param_grid, cv=3, scoring='accuracy')

grid_search.fit(trainEmb, y_train)


resultados = pd.DataFrame(grid_search.cv_results_)


for mean, std, params in zip(
    resultados['mean_test_score'],
    resultados['std_test_score'],
    resultados['params']
):
  print(f"{mean:.4f} (+/-{std:.4f}) para {params}")
'''

'\nparam_grid = [\n    {\n        \'solver\': [\'liblinear\'],\n        \'penalty\': [\'l1\', \'l2\'],\n        \'C\': [0.1, 0.5, 1.0, 2.0, 4.0],\n        \'max_iter\': [50, 100]\n    },\n    {\n        \'solver\': [\'lbfgs\'],\n        \'penalty\': [\'l2\'],\n        \'C\': [0.1, 0.5, 1.0, 2.0, 4.0],\n        \'max_iter\': [50, 100]\n    }\n]\n\n\nmodeloLREmb2 = LogisticRegression()\ngrid_search = GridSearchCV(estimator=modeloLREmb2, param_grid=param_grid, cv=3, scoring=\'accuracy\')\n\ngrid_search.fit(trainEmb, y_train)\n\n\nresultados = pd.DataFrame(grid_search.cv_results_)\n\n\nfor mean, std, params in zip(\n    resultados[\'mean_test_score\'],\n    resultados[\'std_test_score\'],\n    resultados[\'params\']\n):\n  print(f"{mean:.4f} (+/-{std:.4f}) para {params}")\n'

In [ ]:
# REGRESIÓN LOGÍSTICA:

# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********


modeloLREmbFinal = LogisticRegression(solver='liblinear', penalty='l2' , max_iter=100, C=2.0)
modeloLREmbFinal.fit(trainEmb, y_train)


y_pred_val_lr = modeloLREmbFinal.predict(valEmb)

print('LR: Train-accuracy: %.2f%%' % (100*modeloLREmbFinal.score(trainEmb, y_train)))
print('LR: Val-accuracy: %2.f%%' % (100*modeloLREmbFinal.score(valEmb, y_val)))
print('----------------')


print(classification_report(y_val, y_pred_val_lr))
# *********** Aquí termina la sección de agregar código *************


LR: Train-accuracy: 87.57%
LR: Val-accuracy: 88%
----------------
              precision    recall  f1-score   support

           0       0.86      0.89      0.87       216
           1       0.89      0.86      0.88       234

    accuracy                           0.88       450
   macro avg       0.88      0.88      0.88       450
weighted avg       0.88      0.88      0.88       450



In [ ]:
## Para iterar por fors por algunas posibilidades con Random Forest
'''
n_estimators_list = [100, 200]
max_depth_list = [10, 20, None]
min_samples_split_list = [2, 5]
min_samples_leaf_list = [1, 2]
max_features_list = ['sqrt', 'log2']
bootstrap_list = [True, False]
resultados = []

for a in range(len(n_estimators_list)):
  for b in range(len(max_depth_list)):
    for c in range(len(min_samples_split_list)):
      for d in range(len(min_samples_leaf_list)):
        for e in range(len(max_features_list)):
          for f in range(len(bootstrap_list)):
            modeloLREmb = RandomForestClassifier(n_estimators=n_estimators_list[a], max_depth=max_depth_list[b], min_samples_split=min_samples_split_list[c],
                                                 min_samples_leaf=min_samples_leaf_list[d], max_features=max_features_list[e], bootstrap=bootstrap_list[f])
            modeloLREmb.fit(trainEmb, y_train)

            resultados.append({
                'n_estimators': n_estimators_list[a],
                'max_depth': max_depth_list[b],
                'min_samples_split': min_samples_split_list[c],
                'min_samples_leaf': min_samples_leaf_list[d],
                'max_features': max_features_list[e],
                'bootstrap': bootstrap_list[f],
                'train_accuracy': (100*modeloLREmb.score(trainEmb, y_train)),
                'val_accuracy': (100*modeloLREmb.score(valEmb, y_val)),
                'train_accuracy - val_accuracy': (100*modeloLREmb.score(trainEmb, y_train)) - (100*modeloLREmb.score(valEmb, y_val))
            })

df_resultados = pd.DataFrame(resultados)
pd.set_option('display.max_rows', 100)
df_resultados
pd.set_option('display.max_rows', 10)
'''


"\nn_estimators_list = [100, 200]\nmax_depth_list = [10, 20, None]\nmin_samples_split_list = [2, 5]\nmin_samples_leaf_list = [1, 2]\nmax_features_list = ['sqrt', 'log2']\nbootstrap_list = [True, False]\nresultados = []\n\nfor a in range(len(n_estimators_list)):\n  for b in range(len(max_depth_list)):\n    for c in range(len(min_samples_split_list)):\n      for d in range(len(min_samples_leaf_list)):\n        for e in range(len(max_features_list)):\n          for f in range(len(bootstrap_list)):\n            modeloLREmb = RandomForestClassifier(n_estimators=n_estimators_list[a], max_depth=max_depth_list[b], min_samples_split=min_samples_split_list[c],\n                                                 min_samples_leaf=min_samples_leaf_list[d], max_features=max_features_list[e], bootstrap=bootstrap_list[f])\n            modeloLREmb.fit(trainEmb, y_train)\n\n            resultados.append({\n                'n_estimators': n_estimators_list[a],\n                'max_depth': max_depth_list[b

In [ ]:
### GridSearchCV para la busqueda hiperparametros de RF ########
'''
param_grid = {
    'n_estimators': [100, 200],
    'max_depth': [10, 20, None],
    'min_samples_split': [2, 5],
    'min_samples_leaf': [1, 2],
    'max_features': ['sqrt', 'log2'],
    'bootstrap': [True, False]
}


modeloRFEmb2 = RandomForestClassifier()
grid_search = GridSearchCV(estimator=modeloRFEmb2, param_grid=param_grid, cv=3, scoring='accuracy')
grid_search.fit(trainEmb, y_train)


resultados = pd.DataFrame(grid_search.cv_results_)


for mean, std, params in zip(
    resultados['mean_test_score'],
    resultados['std_test_score'],
    resultados['params']
):
  print(f"{mean:.4f} (+/-{std:.4f}) para {params}")
'''

'\nparam_grid = {\n    \'n_estimators\': [100, 200],\n    \'max_depth\': [10, 20, None],\n    \'min_samples_split\': [2, 5],\n    \'min_samples_leaf\': [1, 2],\n    \'max_features\': [\'sqrt\', \'log2\'],\n    \'bootstrap\': [True, False]\n}\n\n\nmodeloRFEmb2 = RandomForestClassifier()\ngrid_search = GridSearchCV(estimator=modeloRFEmb2, param_grid=param_grid, cv=3, scoring=\'accuracy\')\ngrid_search.fit(trainEmb, y_train)\n\n\nresultados = pd.DataFrame(grid_search.cv_results_)\n\n\nfor mean, std, params in zip(\n    resultados[\'mean_test_score\'],\n    resultados[\'std_test_score\'],\n    resultados[\'params\']\n):\n  print(f"{mean:.4f} (+/-{std:.4f}) para {params}")\n'

In [ ]:
# BOSQUE ALEATORIO (Random Forest):

# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********


modeloRFEmbFinal = RandomForestClassifier(n_estimators=100, max_depth=None, min_samples_leaf=2, min_samples_split=2, max_features='sqrt', bootstrap=False)
modeloRFEmbFinal.fit(trainEmb, y_train)

y_pred_val_rf = modeloRFEmbFinal.predict(valEmb)

print('\nRF: Train-accuracy: %.2f%%' % (100*modeloRFEmbFinal.score(trainEmb, y_train)))
print('RF: Val-accuracy: %.2f%%' % (100*modeloRFEmbFinal.score(valEmb, y_val)))
print('----------------')

print(classification_report(y_val, y_pred_val_rf))

# *********** Aquí termina la sección de agregar código *************


RF: Train-accuracy: 100.00%
RF: Val-accuracy: 85.78%
----------------
              precision    recall  f1-score   support

           0       0.84      0.87      0.85       216
           1       0.88      0.85      0.86       234

    accuracy                           0.86       450
   macro avg       0.86      0.86      0.86       450
weighted avg       0.86      0.86      0.86       450



# **Pregunta - 9:**



Reporte del mejor modelo con el conjunto de Prueba (Test).


In [ ]:
# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********


mejor_modelo_emb = modeloLREmbFinal

print('Test-accuracy con el mejor modelo %.2f%%' % (100*mejor_modelo_emb.score(testEmb, y_test)))

pred = mejor_modelo_emb.predict(testEmb)
print('\nMatriz de confusión con el mejor modelo:')
print(confusion_matrix(y_test, pred, labels=[0,1]))

print('\nMatriz de confusión con el mejor modelo en proporciones:')
print(confusion_matrix(y_test, pred, labels=[0,1]) / pred.shape[0])


print('\nClassification_report:')
print(classification_report(y_test, pred))

# *********** Aquí termina la sección de agregar código *************

Test-accuracy con el mejor modelo 84.67%

Matriz de confusión con el mejor modelo:
[[184  32]
 [ 37 197]]

Matriz de confusión con el mejor modelo en proporciones:
[[0.40888889 0.07111111]
 [0.08222222 0.43777778]]

Classification_report:
              precision    recall  f1-score   support

           0       0.83      0.85      0.84       216
           1       0.86      0.84      0.85       234

    accuracy                           0.85       450
   macro avg       0.85      0.85      0.85       450
weighted avg       0.85      0.85      0.85       450



# **Pregunta - 10:**

In [ ]:
# Incluye todas las líneas de código y celdas que consideres adecuadas para este ejercicio.


whole_comments = {}

'''
for comment in X:
  #print(comment)

  response = client.embeddings.create(
      input=comment,
      model="text-embedding-3-small"
  )

  whole_comments[comment] = response.data[0].embedding
'''




In [ ]:
whole_comments_file_name = 'whole_comments.pkl'

**Los Tokens que se usaron de OpenAI al obtener los vectores de los 3000 comentarios originales fueron 43804 con el modelo: text-embedding-3-small**


In [ ]:
#len(whole_comments["Works great."])

1536

In [ ]:
#Se guardan los comentarios
'''
with open('/content/drive/MyDrive/Colab Notebooks/NLP/Actividades/ActividadSemana5_WordEmbeddings/'+whole_comments_file_name, 'wb') as f:
  pickle.dump(whole_comments, f)
'''

In [ ]:
#Se leen los comentarios
with open('/content/drive/MyDrive/Colab Notebooks/NLP/Actividades/ActividadSemana5_WordEmbeddings/'+whole_comments_file_name, 'rb') as f:
    embededd_comments_guardados = pickle.load(f)

In [ ]:
x_train_comments, x_val_and_test_comments, y_train_comments, y_val_and_test_comments = train_test_split(X, Y, train_size=.70, shuffle=True, random_state=1)
x_val_comments, x_test_comments, y_val_comments, y_test_comments = train_test_split(x_val_and_test_comments, y_val_and_test_comments, test_size=.50, shuffle=True, random_state=17)


# verificemos las dimensiones obtenidas:

print('X,y Train:', len(x_train_comments), len(y_train_comments))
print('X,y Val:', len(x_val_comments), len(y_val_comments))
print('X,y Test', len(x_test_comments), len(y_test_comments))

X,y Train: 2100 2100
X,y Val: 450 450
X,y Test 450 450


In [ ]:
#Se genera en np array con los vectores embebidos de cada data set
dimensiones_vectores = 1536


#TRAIN
trainEmbComments = []
for comment in x_train_comments:
    if comment in embededd_comments_guardados:
      trainEmbComments.append(embededd_comments_guardados[comment])
      #print('Si estuvo' + comment)
    else:
      trainEmbComments.append(np.zeros(dimensiones_vectores))
      #print('No estuvo' + comment)

# Se convierte la lista a un array de Numpy
trainEmbComments = np.array(trainEmbComments)

#VAL
valEmbComments = []
for comment in x_val_comments:
    if comment in embededd_comments_guardados:
      valEmbComments.append(embededd_comments_guardados[comment])
    else:
      valEmbComments.append(np.zeros(dimensiones_vectores))

# Se convierte la lista a un array de Numpy
valEmbComments = np.array(valEmbComments)

#TEST
testEmbComments = []
for comment in x_test_comments:
    if comment in embededd_comments_guardados:
      testEmbComments.append(embededd_comments_guardados[comment])
    else:
      testEmbComments.append(np.zeros(dimensiones_vectores))

# Se convierte la lista a un array de Numpy
testEmbComments = np.array(testEmbComments)

In [ ]:
solver_list = ['lbfgs', 'liblinear']
penalty_list = ['l1', 'l2']
max_iter_list = [50, 100]
C_list = [0.1, 0.5, 1.0, 2.0, 4.0, 6.0, 8.0, 10.0]
solver_penalty_combinatons = ['lbfgs-l2', 'lbfgs-None', 'liblinear-l1', 'liblinear-l2']
resultados = []

for a in range(len(solver_list)):
  for b in range(len(penalty_list)):
    for c in range(len(max_iter_list)):
      for d in range(len(C_list)):
        if(solver_list[a]+'-'+penalty_list[b] in solver_penalty_combinatons):
          modeloLREmb_10 = LogisticRegression(solver=solver_list[a], penalty=penalty_list[b], max_iter=max_iter_list[c], C=C_list[d])
          modeloLREmb_10.fit(trainEmbComments, y_train_comments)

          resultados.append({
              'solver': solver_list[a],
              'penalty': penalty_list[b],
              'max_iter': max_iter_list[c],
              'C': C_list[d],
              'train_accuracy': (100*modeloLREmb_10.score(trainEmbComments, y_train_comments)),
              'val_accuracy': (100*modeloLREmb_10.score(valEmbComments, y_val_comments)),
              'train_accuracy - val_accuracy': (100*modeloLREmb_10.score(trainEmbComments, y_train_comments)) - (100*modeloLREmb_10.score(valEmbComments, y_val_comments))
          })

df_resultados = pd.DataFrame(resultados)

df_resultados

solver penalty  max_iter     C  train_accuracy  val_accuracy  \
0       lbfgs      l2        50   0.1       98.000000     96.444444   
1       lbfgs      l2        50   0.5       98.238095     96.888889   
2       lbfgs      l2        50   1.0       98.285714     97.111111   
3       lbfgs      l2        50   2.0       98.619048     97.333333   
4       lbfgs      l2        50   4.0       99.095238     97.333333   
5       lbfgs      l2        50   6.0       99.095238     97.555556   
6       lbfgs      l2        50   8.0       99.142857     97.333333   
7       lbfgs      l2        50  10.0       99.333333     97.333333   
8       lbfgs      l2       100   0.1       98.000000     96.444444   
9       lbfgs      l2       100   0.5       98.238095     96.888889   
10      lbfgs      l2       100   1.0       98.285714     97.111111   
11      lbfgs      l2       100   2.0       98.619048     97.333333   
12      lbfgs      l2       100   4.0       99.095238     97.333333   
13      lbfgs      l2       100   6.0       99.095238     97.555556   
14      lbfgs      l2       100   8.0       99.142857     97.333333   
15      lbfgs      l2       100  10.0       99.333333     97.333333   
16  liblinear      l1        50   0.1       88.952381     90.666667   
17  liblinear      l1        50   0.5       97.523810     96.666667   
18  liblinear      l1        50   1.0       98.000000     96.444444   
19  liblinear      l1        50   2.0       98.428571     96.666667   
20  liblinear      l1        50   4.0       99.142857     97.333333   
21  liblinear      l1        50   6.0       99.285714     97.111111   
22  liblinear      l1        50   8.0       99.476190     96.666667   
23  liblinear      l1        50  10.0       99.571429     97.111111   
24  liblinear      l1       100   0.1       88.952381     90.666667   
25  liblinear      l1       100   0.5       97.523810     96.666667   
26  liblinear      l1       100   1.0       98.000000     96.444444   
27  liblinear      l1       100   2.0       98.428571     96.444444   
28  liblinear      l1       100   4.0       99.142857     97.333333   
29  liblinear      l1       100   6.0       99.285714     97.111111   
30  liblinear      l1       100   8.0       99.476190     96.666667   
31  liblinear      l1       100  10.0       99.571429     97.111111   
32  liblinear      l2        50   0.1       98.000000     96.444444   
33  liblinear      l2        50   0.5       98.238095     96.888889   
34  liblinear      l2        50   1.0       98.285714     97.111111   
35  liblinear      l2        50   2.0       98.761905     97.555556   
36  liblinear      l2        50   4.0       99.095238     97.333333   
37  liblinear      l2        50   6.0       99.095238     97.555556   
38  liblinear      l2        50   8.0       99.095238     97.333333   
39  liblinear      l2        50  10.0       99.285714     97.333333   
40  liblinear      l2       100   0.1       98.000000     96.444444   
41  liblinear      l2       100   0.5       98.238095     96.888889   
42  liblinear      l2       100   1.0       98.285714     97.111111   
43  liblinear      l2       100   2.0       98.761905     97.555556   
44  liblinear      l2       100   4.0       99.095238     97.333333   
45  liblinear      l2       100   6.0       99.095238     97.555556   
46  liblinear      l2       100   8.0       99.095238     97.333333   
47  liblinear      l2       100  10.0       99.285714     97.333333   

    train_accuracy - val_accuracy  
0                        1.555556  
1                        1.349206  
2                        1.174603  
3                        1.285714  
4                        1.761905  
5                        1.539683  
6                        1.809524  
7                        2.000000  
8                        1.555556  
9                        1.349206  
10                       1.174603  
11                       1.285714  
12                       1.761905  
13                     

In [ ]:
modeloLREmb_10_final = LogisticRegression(solver='liblinear', penalty='l1' , max_iter=100, C=0.5)
modeloLREmb_10_final.fit(trainEmbComments, y_train_comments)


y_pred_val_comments_lr = modeloLREmb_10_final.predict(valEmbComments)

print('LR: Train-accuracy: %.2f%%' % (100*modeloLREmb_10_final.score(trainEmbComments, y_train_comments)))
print('LR: Val-accuracy: %2.f%%' % (100*modeloLREmb_10_final.score(valEmbComments, y_val_comments)))
print('----------------')


print(classification_report(y_val_comments, y_pred_val_comments_lr))

LR: Train-accuracy: 97.52%
LR: Val-accuracy: 97%
----------------
              precision    recall  f1-score   support

           0       0.98      0.95      0.96       216
           1       0.96      0.98      0.97       234

    accuracy                           0.97       450
   macro avg       0.97      0.97      0.97       450
weighted avg       0.97      0.97      0.97       450



In [ ]:
n_estimators_list = [100, 200]
max_depth_list = [10, 20, None]
min_samples_split_list = [2, 5]
min_samples_leaf_list = [1, 2]
max_features_list = ['sqrt', 'log2']
bootstrap_list = [True, False]
resultados = []

for a in range(len(n_estimators_list)):
  for b in range(len(max_depth_list)):
    for c in range(len(min_samples_split_list)):
      for d in range(len(min_samples_leaf_list)):
        for e in range(len(max_features_list)):
          for f in range(len(bootstrap_list)):
            modeloLREmb = RandomForestClassifier(n_estimators=n_estimators_list[a], max_depth=max_depth_list[b], min_samples_split=min_samples_split_list[c],
                                                 min_samples_leaf=min_samples_leaf_list[d], max_features=max_features_list[e], bootstrap=bootstrap_list[f])
            modeloLREmb.fit(trainEmbComments, y_train_comments)

            resultados.append({
                'n_estimators': n_estimators_list[a],
                'max_depth': max_depth_list[b],
                'min_samples_split': min_samples_split_list[c],
                'min_samples_leaf': min_samples_leaf_list[d],
                'max_features': max_features_list[e],
                'bootstrap': bootstrap_list[f],
                'train_accuracy': (100*modeloLREmb.score(trainEmbComments, y_train_comments)),
                'val_accuracy': (100*modeloLREmb.score(valEmbComments, y_val_comments)),
                'train_accuracy - val_accuracy': (100*modeloLREmb.score(trainEmbComments, y_train_comments)) - (100*modeloLREmb.score(valEmbComments, y_val_comments))
            })

df_resultados = pd.DataFrame(resultados)
pd.set_option('display.max_rows', 100)
df_resultados


n_estimators  max_depth  min_samples_split  min_samples_leaf max_features  \
0            100       10.0                  2                 1         sqrt   
1            100       10.0                  2                 1         sqrt   
2            100       10.0                  2                 1         log2   
3            100       10.0                  2                 1         log2   
4            100       10.0                  2                 2         sqrt   
5            100       10.0                  2                 2         sqrt   
6            100       10.0                  2                 2         log2   
7            100       10.0                  2                 2         log2   
8            100       10.0                  5                 1         sqrt   
9            100       10.0                  5                 1         sqrt   
10           100       10.0                  5                 1         log2   
11           100       10.0                  5                 1         log2   
12           100       10.0                  5                 2         sqrt   
13           100       10.0                  5                 2         sqrt   
14           100       10.0                  5                 2         log2   
15           100       10.0                  5                 2         log2   
16           100       20.0                  2                 1         sqrt   
17           100       20.0                  2                 1         sqrt   
18           100       20.0                  2                 1         log2   
19           100       20.0                  2                 1         log2   
20           100       20.0                  2                 2         sqrt   
21           100       20.0                  2                 2         sqrt   
22           100       20.0                  2                 2         log2   
23           100       20.0                  2                 2         log2   
24           100       20.0                  5                 1         sqrt   
25           100       20.0                  5                 1         sqrt   
26           100       20.0                  5                 1         log2   
27           100       20.0                  5                 1         log2   
28           100       20.0                  5                 2         sqrt   
29           100       20.0                  5                 2         sqrt   
30           100       20.0                  5                 2         log2   
31           100       20.0                  5                 2         log2   
32           100        NaN                  2                 1         sqrt   
33           100        NaN                  2                 1         sqrt   
34           100        NaN                  2                 1         log2   
35           100        NaN                  2                 1         log2   
36           100        NaN                  2                 2         sqrt   
37           100        NaN                  2                 2         sqrt   
38           100        NaN                  2                 2         log2   
39           100        NaN                  2                 2         log2   
40           100        NaN                  5                 1         sqrt   
41           100        NaN                  5                 1         sqrt   
42           100        NaN                  5                 1         log2   
43           100        NaN                  5                 1         log2   
44           100        NaN                  5                 2         sqrt   
45           100        NaN                  5                 2         sqrt   
46           100        NaN                  5                 2         log2   
47           100        NaN                  5                 2         log2   
48           200       10.0        

In [ ]:
pd.set_option('display.max_rows', 10)

In [ ]:
modeloRFEmb_10_final = RandomForestClassifier(n_estimators=100, max_depth=10, min_samples_split=2, min_samples_leaf=2, max_features='sqrt', bootstrap=True)
modeloRFEmb_10_final.fit(trainEmbComments, y_train_comments)

y_pred_val_comments_rf = modeloRFEmb_10_final.predict(valEmbComments)

print('\nRF: Train-accuracy: %.2f%%' % (100*modeloRFEmb_10_final.score(trainEmbComments, y_train_comments)))
print('RF: Val-accuracy: %.2f%%' % (100*modeloRFEmb_10_final.score(valEmbComments, y_val_comments)))
print('----------------')

print(classification_report(y_val_comments, y_pred_val_comments_rf))


RF: Train-accuracy: 99.90%
RF: Val-accuracy: 96.00%
----------------
              precision    recall  f1-score   support

           0       0.96      0.95      0.96       216
           1       0.96      0.97      0.96       234

    accuracy                           0.96       450
   macro avg       0.96      0.96      0.96       450
weighted avg       0.96      0.96      0.96       450



In [ ]:
mejor_modelo_emb_comments = modeloLREmb_10_final

print('Test-accuracy con el mejor modelo %.2f%%' % (100*mejor_modelo_emb.score(testEmbComments, y_test_comments)))

pred = mejor_modelo_emb_comments.predict(testEmbComments)
print('\nMatriz de confusión con el mejor modelo:')
print(confusion_matrix(y_test_comments, pred, labels=[0,1]))

print('\nMatriz de confusión con el mejor modelo en proporciones:')
print(confusion_matrix(y_test_comments, pred, labels=[0,1]) / pred.shape[0])


print('\nClassification_report:')
print(classification_report(y_test_comments, pred))

Test-accuracy con el mejor modelo 96.44%

Matriz de confusión con el mejor modelo:
[[212   4]
 [ 13 221]]

Matriz de confusión con el mejor modelo en proporciones:
[[0.47111111 0.00888889]
 [0.02888889 0.49111111]]

Classification_report:
              precision    recall  f1-score   support

           0       0.94      0.98      0.96       216
           1       0.98      0.94      0.96       234

    accuracy                           0.96       450
   macro avg       0.96      0.96      0.96       450
weighted avg       0.96      0.96      0.96       450



# **Pregunta - 11:**



Incluye tus comentarios finales de la actividad.

### ++++++++ Inicia la sección de agregar texto: +++++++++++

None

### ++++++++ Termina la sección de agregar texto: +++++++++++

# **Fin de la Actividad de Vectores Embebidos - OpenAI**